xarray/core/treenode.py

```diff
@@ -25,7 +21,7 @@ class InvalidTreeError(Exception):
 
 
 class NotFoundInTreeError(ValueError):
-    """Raised when operation can't be completed because one node is part of the expected tree."""
+    """Raised when operation can't be completed because one node is not part of the expected tree."""
 
 
 class NodePath(PurePosixPath):
```

Comment:
Are we exposing `NodePath`? I suggested it, but after talking with @etienneschalk decided maybe it wasn't particularly helpful https://github.com/xarray-contrib/datatree/issues/205.

What's the logic for making it a pure path? The posix path is to ensure the slashes are in a consistent direction (remember this isn't a real filepath so shouldn't change automatically on windows systems).

Reply:
I am not sure about this, but I believe the difference is that `PurePosixPath` doesn't have methods that require read/write access (like `touch`, `remove`, `mkdir`, `iterdir`, ...), while `PosixPath` does.

The main difference between filesystems and `datatree` is that there's only a single filesystem instance at a time, while you can have multiple `DataTree` objects at the same time. This means that a `NodePath` inheriting from `PosixPath` would have to be created from the `DataTree` object and would require a lot of discipline to not be confusing.

On the other hand, a `NodePath` that inherits from `PurePosixPath` would help a bit with path manipulations (`name`, `root`, `joinpath()`, `parts`, `parent`, `parents`, ...) that we would otherwise have to use `posixpath` for, but I didn't check if all of these exist in that module, and they might also not be as convenient to use.

Reply:
Sorry, I wrote my above comment too quickly - we both agree it should stay as a pure path.

We could also consider adding extra methods to this `NodePath` class if it simplifies other parts of the code.

Reply:
Related issue: https://github.com/xarray-contrib/datatree/issues/283 (Consistency between DataTree methods and pathlib.PurePath methods)

To me the graal would be to have datatree more and more compatible with `PurePosixPath` (what `NodePath` already does by inheriting from it, but we can imagine having an API accepting both strings and `PurePosixPath`, without the need to expose `NodePath` publicly).

As @keewis mentioned , there is a single instance of the filesystem whereas there can be multiple DataTrees. So a single DataTree could act like a "small filesystem on its own" and  implement the `PosixPath` methods. 

```
NodePath ~ PurePosixPath
DataTree ~ PosixPath 
``` 

The idea here is to delegate to `pathlib`  the diffcult task of choosing the best names for methods that help manipulating trees. For an audience of developers that are already used to writing scripts with `pathlib`, this would make working with datatree automatic! And this also gives plenty of ideas of features. Also, with hierarchical-formats such as Zarr that rely on the filesystem, openable by datatree, the proximity is innate.

Concrete example: working with multi-resolution rasters (a usecase of datatree). By using f-strings and glob, you can write resolution independant code once, like you would have done by manipulating a data structure on the file system

The schema on the [pathlib](https://docs.python.org/3/library/pathlib.html) doc (arrows are inheritance):

```mermaid
flowchart BT
  PurePosixPath --> PurePath
  Path --> PurePath
  PureWindowsPath --> PurePath
  PosixPath --> PurePosixPath
  PosixPath --> Path
  WindowsPath --> PureWindowsPath
  WindowsPath --> Path
```

How I see it:

```mermaid
flowchart BT
subgraph Pure
  PurePosixPath --> PurePath
  NodePath --> PurePosixPath
  end
subgraph Concrete
  DataTree --> NodePath
  PosixPath --> PurePosixPath
  end
```

(I removed `Path` for simplification)

I am not 100% sure it fully makes sense... The nuance is that when you instanciate a Path, it is always implicit that it is bound to the filesystem, while each DataTree instantiation creates its own data representation.

Reply:
Thank you @etienneschalk for all your thoughts here!

> So a single DataTree could act like a "small filesystem on its own" and implement the PosixPath methods.
>
> I am not 100% sure it fully makes sense... The nuance is that when you instanciate a Path, it is always implicit that it is bound to the filesystem, while each DataTree instantiation creates its own data representation.

I strongly feel that DataTree / NodePath objects should not be associated with concrete filesystem paths.

However this

> To me the graal would be to have datatree more and more compatible with PurePosixPath (what NodePath already does by inheriting from it, but we can imagine having an API accepting both strings and PurePosixPath, without the need to expose NodePath publicly).

seems like a reasonable idea, and I think we should continue the discussion of that in https://github.com/xarray-contrib/datatree/issues/283.

For now I don't think there is anything we need to do in this PR specifically - we can revisit the API choices / accepting path-like types in a future PR.